text/0000-thread-affinity.md

```diff
@@ -0,0 +1,133 @@
+    let cpu_id = cpu_lock.cpu_id();
+    // thread logic goes here
+    // the cpu_lock will unlock from the core once out of scope
+});
+```
+
+Here's the equivalent but with some syntatical sugar to hide the lock if the user doesn't care to access it:
+
+```
+let t = thread::lock(move || {
+    // thread logic goes here
+    // the cpu_lock will unlock from the CPU core once out of scope
+});
+```
+
+The user may want more fine grained control of which core it should be locked to.
+
+### Locking to a specific CPU core
+
+```
+// locks thread to cpu core 2
+let t = thread::spawn(move || {
+    let cpu_no = 2;
+    let cpu_lock = CpuLock::lock_on(cpu_no).unwrap(); // the current thread is now locked to an arbitrary free CPU core
+    // thread logic goes here
+    // the cpu_lock will unlock from the core once out of scope
+});
+```
+
+```
+// locks thread to cpu core 2
+let cpu_no = 2;
+let t = thread::lock_on(2,move || {
+    let cpu_lock = CpuLock::lock_on(cpu_no).unwrap(); // the current thread is now locked to an arbitrary free CPU core
+    let cpu_id = cpu_lock.cpu_id();
+    // thread logic goes here
+    // the cpu_lock will unlock from the core once out of scope
+});
+```
+
+### Unlocking prematurely
+
+```
+// locks thread to cpu core 2
+let t = thread::spawn(move || {
+    let cpu_lock = CpuLock::lock().unwrap(); // the current thread is now locked to an arbitrary free CPU core    
+    // thread logic goes here
+    cpu_lock.unlock();
+    // other thead logic - like cleanup, etc.
+});
+```
+
+## Implementation
+
+The implementation is quite simple as it just needs to store the underlying thread's id and invoke system calls to lock/unlock.
+
+*  Windows -> *SetThreadAffinityMask*
+
+*  Nix: -> *PTHREAD_SETAFFINITY_NP(3)*
+
+The implementation needs to know the number of available CPUS and manage state to know which CPUs are available for arbitrary locking.
+
+To get the affinity mask from the cpu no is trivial:
+
+```
+let cpu_no = 2;
+let affinity_mask = 1 << cpu_no;
+```
+
+### Corner-cases
+
+1. Passing an cpu number out of bounds.
+
+2. Moving the lock out of the thread.
+
+3. Moving the locked from one thread to another thread.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I believe the drawbacks of the current RFC are:
+
+1. This RFC does not provide any empirical evidence of the computational benefits. I can spend time to provide benchmarks from a C/C++ benchmark.
+2. Thread affinity can have the adverse effect - degrading performance if it interferes with the OS's scheduler if used in the wrong way.
+3. My lack of experience of Rust could be prohibiting an optimal design.
+4. Could be regarded as adding complexity to the standard library.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Calling unsafe OS-specific system calls to create threads and changing the affinity mask because `std::thread::Thread` do not expose the underling OS-specific thread id (AFIAW).
```

Comment:
`JoinHandle` is what actually contains the OS-specific thread handle. On Windows `JoinHandle` implements `AsRawHandle` and `IntoRawHandle` and on non-windows it implements `JoinHandleExt`. Also getting the handle to the current thread is pretty simple (`GetCurrentThread` or `pthread_self` depending on your platform).

https://github.com/rust-lang/rust/pull/30177